### PR TITLE
Expose HumandHandle as a structure

### PR DESCRIPTION
**Pull request by FirelightFlagboy**
Depends #5052

### CODE REVIEW

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: bindings/generator/generate.py [hunk 1]
```diff
@@ -217,6 +217,14 @@ class StructSpec(BaseTypeInUse):
     kind = "struct"
     name: str
     attributes: OrderedDict[str, BaseTypeInUse]
+    getters: dict[str, str]
+    init: dict[str, str] | None
+
+    def get_value(self, attr_name: str) -> str | None:
+        value = self.getters.get(attr_name, None)
+        if value is not None:
+            return value + "()"
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/generate.py:226]: I'm not sure what you mean by that, that method is called in the jinja template.

did you mean that the code present in the method should be put in the template ?

File: bindings/generator/generate.py [hunk 2]
```diff
@@ -217,6 +217,14 @@ class StructSpec(BaseTypeInUse):
     kind = "struct"
     name: str
     attributes: OrderedDict[str, BaseTypeInUse]
+    getters: dict[str, str]
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/generate.py]: done

File: bindings/generator/templates/binding_electron_meths.rs.j2 [hunk 3]
```diff
@@ -344,11 +344,15 @@ fn {{ struct_js_to_rs_function_name(struct) }}<'a>(
         {{ render_downcasted_js_to_rs("js_val", attr_type, mut_cx_ref="cx") | indent(8) }}
     };
 {% endfor %}
+{% if struct.init is none %}
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/templates/binding_electron_meths.rs.j2]: done

File: bindings/generator/api/common.py [hunk 4]
```diff
@@ -109,8 +109,14 @@ class DeviceLabel(StrBasedType):
     pass
 
 
-class HumanHandle(StrBasedType):
-    pass
+class HumanHandle(Structure):
+    email: str
+    label: str
+    custom_getters: dict[str, str] = {"email": "email", "label": "label"}
+    custom_init: dict[str, str] = {
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/api/common.py:116]: done

File: bindings/generator/api/common.py [hunk 5]
```diff
@@ -109,8 +109,14 @@ class DeviceLabel(StrBasedType):
     pass
 
 
-class HumanHandle(StrBasedType):
-    pass
+class HumanHandle(Structure):
+    email: str
+    label: str
+    custom_getters: dict[str, str] = {"email": "email", "label": "label"}
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/api/common.py]: Currently we couldn't specify lifetime parameters in closure:

I've tried to use a closure instead as proposed but it result in the following error:

```
error: lifetime may not live long enough
   --> bindings/web/src/meths.rs:158:63
    |
158 | ... = ((|obj: &libparsec::HumanHandle| -> &str { obj.email() })(&rs_obj)).in...
    |               -                           -      ^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |               |                           |
    |               |                           let's call the lifetime of this reference `'2`
    |               let's call the lifetime of this reference `'1`
```

File: bindings/generator/generate.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply bindings/generator/generate.py]: the template should call the function:
```
{ my_func }()
```

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/generate.py]: ok

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply bindings/generator/generate.py]: The method `get_value` was removed

**Comment by touilleMan** [comment]
I've created a branch implementing the closure-based approach we discussed on slack last week:

https://github.com/Scille/parsec-cloud/tree/expose-HumanHandle-as-struct-electric-boogaloo

(the important commit: https://github.com/Scille/parsec-cloud/commit/8f56cf9cb07547c7023509e136289acc64161275)

The in-elegant function-in-a-closure we were wondering seems pretty small and no big deal in the end:
```rust
|obj| { fn access(obj: &libparsec::HumanHandle) -> &str { obj.email() } access(obj) }
```

What do you think ?

**Comment by touilleMan** [comment]
> It's not as statisfactory as I would like: clippy raise the following warning

You should be able to avoid this warning by doing:
```rust
{
  let custom_getter = |obj| {
    fn access(obj: &libparsec::HumanHandle) -> &str {
      obj.email()
    }
    access(obj)
  };
  custom_getter(&rs_obj)
}
```

(this is what is currently done for `custom_to_rs_string`&co)

**Pull request closed without merging**